### ISSUE TITLE
Template preprocessing(?)

### ISSUE THREAD

**Issue by dfreeman** [salsify/ember-css-modules#1]
Currently to reference a class name from a component/controller's corresponding CSS module, users must insert an explicit binding to a field of the `styles` object in their template.

```hbs
<div class="{{styles.container}}">
  <span class="framework-class {{styles.text}}">Hello</span>
</div>
```

This has the advantage of making it very clear where the class names are coming from, but is fairly verbose. One alternative might be to introduce a dedicated attribute for local classes and then provide a Handlebars preprocessor that would transform the following template, making it equivalent to the above:

```hbs
<div local-class="container">
  <span class="framework-class" local-class="text">Hello</span>
<div>
```

You could even make `class` local by default and instead make global classes the special case:

```hbs
<div class="container">
  <span global-class="framework-class" class="text">Hello</span>
</div>
```

Would introducing such a transform even be a good idea? 

Some cons:
 - It would introduce fairly magical (i.e. nonobvious) behavior
 - Handlebars AST transforms are technically a private API

Some pros:
 - Templates and CSS source more closely align; potentially better readability
 - Easier migration – more existing styles would continue to "just work" with fewer changes

**Comment by cibernox** [comment]
I was thinking about this actually. With a handlebars AST tranformation all interpolations could be done in compile time. 
I'd suggest some syntax like `class="global-class {{someLocalPropertyLikeActiveOrDisabledOrSomething}} {{:style.moduleCSSClass}}"`

I think that it removes the cognitive overhead of having a different attribute that text-editors don't know to autocomplete.

**Comment by dfreeman** [comment]
I actually started down the path of doing template interpolation at compile time (I still have a proof-of-concept branch somewhere), but unfortunately it doesn't mesh well with class composition.

```css
.foo {
  composes: bar;
}
```

Given the above, the CSS Modules compilation process doesn't change the definition of `.foo` to incorporate `.bar`'s styles like a Sass `@include` or add the `.foo` selector to `.bar`'s definition like `@extend` – in fact, the class definitions aren't changed at all.
Instead, the value for `foo` in the exported styles hash will contain references to the namespaced version of both `foo` and `bar`, so any element with`class={{styles.foo}}` will end having both classes assigned.

This means that to statically interpolate the namespaced classes during template compilation, the stylesheets need to already have been processed, and we need access to that output from the AST transform. Ember CLI's build pipeline doesn't guarantee the former, and it actively prevents the latter, so I settled for pulling the class names at runtime.

Given that, I'd be hesitant to introduce a microsyntax like `{{:moduleCSSClass}}`, since it would just be notational shorthand for `{{styles.moduleCSSClass}}` rather than meaning something distinct at build time.

**Comment by cibernox** [comment]
> Given that, I'd be hesitant to introduce a microsyntax like {{:moduleCSSClass}}

No question, this microsyntax was only an idea to mark nodes for a possible precompilation by a handlebars AST. I doesn't makes sense just to save a few keystrokes if the interpolation happens in runtime.

**Comment by dfreeman** [comment]
This is now released as `local-class="foo"` in 0.4.0.
